### PR TITLE
precision time protocol (ptp) encapsulated in UDP

### PR DESCRIPTION
**Pull request by psglinux**
PTP is used for synchronizing time across network.  

Added support for PTP packet encapsulated in UDP packets. Have tested for 5 different types of ptp messages. 

Greetings !!
I wrote the ptp over udp extension for TCP dump. I've raised a pull request for this. Would be looking forward for the review and suggestion for improvements. There was a ptp extension written by @lunn which is for ptp over L2 (wasn't aware about it till a few days back when i was thinking of sending the code 
 upstream and read the README :-) ). If needed i can integrate both these code. 
Please let me know your thought and comments.

### CODE REVIEW

File: print-ptp.c [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+    *len -= 1;
+    bp += 1;
+    /* GM clock accuracy */
+    gm_clk_acc = EXTRACT_U_1(bp);
+    ND_PRINT(", gm clock accuracy : %u", gm_clk_acc);
+    *len -= 1;
+    bp += 1;
+    /* GM clock variance */
+    gm_clk_var = EXTRACT_BE_U_2(bp);
+    ND_PRINT(", gm clock varience : %u", gm_clk_var);
```
Comment (agnosticdev) [inline print-ptp.c]: A small cleanup for `variance`.

File: print-ptp.c [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+    len -= 8; bp += 8; port_id = EXTRACT_BE_U_2(bp); ND_PRINT(", port id : %u", port_id);
+
+    /* sequence ID */
+    len -= 2; bp += 2; seq_id = EXTRACT_BE_U_2(bp); ND_PRINT(", seq id : %u", seq_id);
+
+    /* control */
+    len -= 2; bp += 2; control = EXTRACT_U_1(bp) ;
+    ND_PRINT(", control : %u (%s)", control, tok2str(ptp_msg_type, "none", EXTRACT_U_1(&control)));
+
+    /* log messsage interval */
```
Comment (agnosticdev) [inline print-ptp.c]: A minor cleanup for `message`.

File: print-ptp.c [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+#define PTP_LOGMSG_MASK     0xFF
+
+/* mask based on the flags 2 bytes */
+
+#define PTP_L161_MASK               0x1
+#define PTP_L1_59_MASK              0x2
+#define PTP_UTC_REASONABLE_MASK     0x4
+#define PTP_TIMESCALE_MASK          0x8
+#define PTP_TIME_TRACABLE_MASK      0x10
+#define PTP_FREQUENCY_TRACABLE_MASK 0x20
```
Comment (agnosticdev) [inline print-ptp.c]: Looking at the PTP description: https://support.huawei.com/hedex/pages/EDOC100010596830008125/05/EDOC100010596830008125/05/resources/message/cd_feature_1588v2_format-general.html#cd_feature_eth_5__packetformat

Was this meant to be described as `PTP_TIME_TRACEABLE_MASK` and `PTP_FREQUENCY_TRACEABLE_MASK`

**Comment by mcr** [comment]
You'll need to redo the ND_ macros, as we have newer ones that do a better job.

**Comment by fxlb** [comment]
You should **rebase** you branch on the-tcpdump-group:master.

**Comment by psglinux** [comment]
Thanks for pointing out. I haven't made the necessary changes yet . Will
kepp this in mind before sending the pull request again for this change.

**Comment by fxlb** [comment]
Please, use the new GET_ macros instead of EXTRACT_ ones (See 176e182 and ee68aa3).

**Comment by psglinux** [comment]
@fxlb will take care of the changes to use EXTRACT_ instead of GET_

**Comment by psglinux** [comment]
@mcr you mentioned  "ND_ macros, as we have newer ones that do a better job." in the code ND_PRINT is being used. Are you suggesting i use a different macro for the print ?

**Comment by psglinux** [comment]
My earlier merge was a mess. so please ignore the failed noises. Sorted that out. may be will raise a new pull request, once the comments are clear to me i will address all of them.

**Comment by fxlb** [comment]
There is no need for a new PR, you could:
```
# Save your master branch
git checkout master; git checkout -b master-save; git checkout -
# Cancel the last 2 commits (merge branch 'master' and revert the merge)
git reset --hard HEAD~2
# Configuring the remote for your fork (if not already done)
git remote add upstream https://github.com/the-tcpdump-group/tcpdump.git
# Fetch upstream
git fetch upstream
# Rebase on upstream/master
git rebase upstream/master
# Fix the conflict in tests/TESTLIST, editing the file
git add tests/TESTLIST
git rebase --continue
# Do the GET_ and others updates
# Test
make -s check
# Push (forced)
git push -f
```

**Comment by psglinux** [comment]
Please let me know if there are any other comments or changes that needs to be taken care.

File: print-ptp.c (hunk 2 above)
Comment (fxlb) [reply print-ptp.c]: To do.

File: print-ptp.c (hunk 3 above)
Comment (fxlb) [reply print-ptp.c]: > Please, use the new GET_ macros instead of EXTRACT_ ones

Use the GET_  macros, not the get_ inline functions.

**Comment by fxlb** [comment]
A new rebase is needed.

**Comment by psglinux** [comment]
Please let me know if any thing else is needed for this pull request.

**Comment by fxlb** [comment]
The commit 4ad9e8488cdf431e5d4ad8983f50bd104e36b6e4 uses get_ inline functions instead of the GET_ macros asked (see a previous comment). Please remove it with a `rebase -i` and substitute EXTRACT_ macros by  GET_ ones (like in ee68aa36460d7efeca48747f33b7f2adc0900bfb).

**Comment by psglinux** [comment]
Let me know if the patch is good for merge.

**Comment by fxlb** [comment]
Please, see the indent problems in the  previous comments.

**Comment by fxlb** [comment]
Need to fix "messsage" typo.

**Comment by psglinux** [comment]
Let me know if any thing else needs to be done to get this one merged.

**Comment by fxlb** [comment]
For a better log history, you should `git rebase -i HEAD~7`
With `r, reword` for 050b7beb1c7837941c80eba64856ab8a11be6ebd (see `h) A commit message must have` in CONTRIBUTING file.
For the others commit `s, squash` ot better I think `f, fixup`.

**Comment by fxlb** [comment]
Many indent problems already noted, see [Files changed] tab.

**Comment by psglinux** [comment]
@fxlb have fixed the tabs in the other files that I've changed. New files do not have tabs in them and indented with 4 spaces.

**Comment by fxlb** [comment]
Indent problem:
indent1

**Comment by fxlb** [comment]
Another indent problem:
image

**Comment by fxlb** [comment]
@psglinux 
Merged, Thank you!

**Pull request closed without merging**